commit dbaab49f92ff6ae6255762a948375e4036cbdbd2
Author: Vivek Goyal <vgoyal@in.ibm.com>
Date:   Sat Oct 21 18:37:03 2006 +0200

    [PATCH] x86-64: Overlapping program headers in physical addr space fix
    
    o A recent change to vmlinux.ld.S file broke kexec as now resulting vmlinux
      program headers are overlapping in physical address space.
    
    o Now all the vsyscall related sections are placed after data and after
      that mostly init data sections are placed. To avoid physical overlap
      among phdrs, there are three possible solutions.
    	- Place vsyscall sections also in data phdrs instead of user
    	- move vsyscal sections after init data in bss.
    	- create another phdrs say data.init and move all the sections
    	  after vsyscall into this new phdr.
    
    o This patch implements the third solution.
    
    Signed-off-by: Vivek Goyal <vgoyal@in.ibm.com>
    Signed-off-by: Andi Kleen <ak@suse.de>
    Cc: Magnus Damm <magnus@valinux.co.jp>
    Cc: Andi Kleen <ak@suse.de>
    Cc: "Eric W. Biederman" <ebiederm@xmission.com>
    Signed-off-by: Andrew Morton <akpm@osdl.org>

diff --git a/arch/x86_64/kernel/vmlinux.lds.S b/arch/x86_64/kernel/vmlinux.lds.S
index b9df2ab..1283614 100644
--- a/arch/x86_64/kernel/vmlinux.lds.S
+++ b/arch/x86_64/kernel/vmlinux.lds.S
@@ -17,6 +17,7 @@ PHDRS {
 	text PT_LOAD FLAGS(5);	/* R_E */
 	data PT_LOAD FLAGS(7);	/* RWE */
 	user PT_LOAD FLAGS(7);	/* RWE */
+	data.init PT_LOAD FLAGS(7);	/* RWE */
 	note PT_NOTE FLAGS(4);	/* R__ */
 }
 SECTIONS
@@ -131,7 +132,7 @@ SECTIONS
   . = ALIGN(8192);		/* init_task */
   .data.init_task : AT(ADDR(.data.init_task) - LOAD_OFFSET) {
 	*(.data.init_task)
-  } :data
+  }:data.init
 
   . = ALIGN(4096);
   .data.page_aligned : AT(ADDR(.data.page_aligned) - LOAD_OFFSET) {
